### DISCUSSION TITLE
Huge temporal files for write operation on relatively big parquet dataset

### DISCUSSION THREAD

**Discussion by Mmoncadaisla** [duckdb/duckdb#12358] [Q&A]
### Question

I have a relatively large table stored as a set of `.parquet` files in a Google Cloud Storage bucket (~180 GB in total), these are individual and "independent" files that I'm trying to reorganise, hoping that it will be possible to then query the table in a more performant way.

```bash
gsutil du -sh gs://bucket_name/path/to/esa_worldcover_10m_v200/
178.14 GiB   gs://bucket_name/path/to/esa_worldcover_10m_v200
```

In order not to hit OOMs, I'm using the `temp_directory` setting to allow DuckDB to spill to disk when needed, the issue is that the space required seems to be way bigger than I had imagined and I'm unsure if there's something completely wrong or it is just expected, as it has already taken **3.5 TB** of SSD disk space and the temp data dir continues to grow.

```bash
df -h /tmp/
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda1       5.3T  3.5T  1.7T  68% /
```

#### Environment

- e2-standard-8 VM in Google Compute Engine 32 GB RAM and 8 CPUs
- `duckdb` version `0.10.3`

#### Data source

The original data source is the [10m Annual Land Use Land Cover (9-class) V2 dataset available in the Planetary Computer Hub](https://planetarycomputer.microsoft.com/dataset/io-lulc-annual-v02), this dataset was pulled and transformed into H3 indexes  in chunks at resolution 13 using the [`h3ronpy`](https://h3ronpy.readthedocs.io/en/latest/) and polars libraries, storing the result in `.parquet` files in the GCS bucket.

#### Script

The idea was to improve the structure by partitioning by the H3 index (calculated for a lower resolution) and ordering by both the lower and the highest resolution available.

```python
import duckdb
from fsspec import filesystem

con = duckdb.connect(config={"allow_unsigned_extensions": "true"})

h3_ext_path = '/path/to/h3ext.duckdb_extension'
    
con.install_extension('httpfs')
con.install_extension('spatial')

con.load_extension('httpfs')
con.load_extension('spatial')
con.load_extension(h3_ext_path)


try:
    con.query("""
            CREATE PERSISTENT SECRET (TYPE GCS, KEY_ID 'GOOG1EXXXXXXXXX', SECRET 'XXXXXXXX');
            """)
except Exception as e:
    print(e)


duckdb.register_filesystem(filesystem('gcs'))


query_str = """COPY (
     WITH _data AS (SELECT h3_cell_to_parent(cast(h3 as ubigint), 8) as h3_res8, h3 as h3_res13, value from read_parquet('gs://bucket_name/path/to/esa_worldcover_10m_v200/*.parquet')) SELECT * FROM _data ORDER BY h3_res8, h3_res13
)
TO 'gs://bucket_name/path/to/optimized/esa_worldcover_10m_v200/'
(FORMAT PARQUET, COMPRESSION ZSTD, PARTITION_BY (h3_res8), ROW_GROUP_SIZE 100_000);"""


con.query("""SET temp_directory = '/tmp';""")
con.query("""SET memory_limit = '28GB';""")
con.query("""SET max_temp_directory_size = '5300GB';""")

con.query("""SELECT *
FROM duckdb_settings() where name IN ('temp_directory', 'memory_limit')""")

con.query(query_str)
```

**Comment by Maxxen** [comment]
Have you tried disabling the [preserve insertion order](https://duckdb.org/docs/guides/performance/how_to_tune_workloads.html#the-preserve_insertion_order-option) setting?

**Reply by Mmoncadaisla** [reply]
Thanks for the suggestion @Maxxen!

I started over and followed this advice, this time using duckdb v `1.0.0`. I also thought that perhaps storing the table as-is in the duckdb database first could potentially lead to improved performance. 

```python
import duckdb
from fsspec import filesystem
import logging

# Configure logging
logging.basicConfig(filename="ddb_transform.log", level=logging.INFO)

con = duckdb.connect(config={"allow_unsigned_extensions": "true"})

h3_ext_path = '/path/to/h3ext.duckdb_extension'

# Install and load extensions
try:
    con.install_extension('httpfs')
    con.install_extension('spatial')
    con.load_extension('httpfs')
    con.load_extension('spatial')
    con.load_extension(h3_ext_path)
except Exception as e:
    logging.error(f"Error installing or loading extensions: {e}")

# Register GCS filesystem
duckdb.register_filesystem(filesystem('gcs'))

# Set table name and maximum resolution
table_name = "esa_worldcover_10m_v200"
max_res = 13

# Set DuckDB configuration
logging.info("Setting DuckDB configuration")
con.query("""SET temp_directory = '/tmp';""")
con.query("""SET memory_limit = '28GB';""")
con.query("""SET max_temp_directory_size = '3990GB';""")
con.query("""SET preserve_insertion_order = false;""")
con.query("""SET enable_progress_bar = true;""")

logging.info(con.query("""SELECT * FROM duckdb_settings()"""))

# Create table from Parquet file
create_table_str = f"""CREATE TABLE {table_name} AS (SELECT * FROM read_parquet('gs://bucket_name/path/to/{table_name}/*.parquet'))"""

try:
    logging.info(f"Creating table {table_name}")
    con.query(create_table_str)
    logging.info(con.query(f"""SELECT * FROM {table_name} USING SAMPLE 5;"""))
except Exception as e:
    logging.error(f"Error creating table: {e}")

# Copy data to optimized Parquet table
copy_query_str = f"""COPY (
     WITH _data AS (SELECT h3_cell_to_parent(cast(h3 as ubigint), 8) as h3_res8, h3 as h3_res{max_res}, value from {table_name}) SELECT * FROM _data ORDER BY h3_res8, h3_res{max_res}
)
TO 'gs://bucket_name/path/to/optimized/{table_name}/'
(FORMAT PARQUET, COMPRESSION ZSTD, PARTITION_BY (h3_res8), ROW_GROUP_SIZE 100_000);"""

try:
    logging.info(f"Copying data to optimized table {table_name}")
    con.query(copy_query_str)
except Exception as e:
    logging.error(f"Error copying data: {e}")

# Close connection
con.close()
```

The table has not even been created yet, and it's taken over 1 TB which seemed too much for me given that the original size in .parquet files is ~180 GB, being way different from the 20% that the docs reflect here https://duckdb.org/docs/internals/storage.html#disk-usage

```bash
df -h
Filesystem      Size  Used Avail Use% Mounted on
udev             16G     0   16G   0% /dev
tmpfs           3.2G  476K  3.2G   1% /run
/dev/sda1       4.0T  1.1T  2.8T  28% /
tmpfs            16G     0   16G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
/dev/sda15      124M   12M  113M  10% /boot/efi
tmpfs           3.2G     0  3.2G   0% /run/user/1000
```

**Reply by Mmoncadaisla** [reply]
FYI I'm trying the same workflow against a smaller table using a VM with 256 GB of RAM and it's having difficulties simply creating the table as well:

```bash
gsutil du -sh gs://bucket_name/path/to/alos_dem_30m/
42.4 GiB     gs://bucket_name/path/to/alos_dem_30m
```


```bash
free -m
               total        used        free      shared  buff/cache   available
Mem:          257942      240897        2923           0       15833       17044
Swap:              0           0           0
```

It ended up giving an OOM:

```
ERROR:root:Error creating table: Out of Memory Error: Failed to allocate block of 262144 bytes (bad allocation)
```

**Reply by Maxxen** [reply]
Here's two things to try, separately or together:
- `SET threads=1` - When partitioning each thread works independently and will buffer their own shards of the partitions until the end when everything gets merged all at once.
- Remove the `ORDER BY` as that will also have to buffer and spill to disk all incoming data before it can proceed with the next part of the pipeline.

**Reply by Mmoncadaisla** [reply]
I will test both of these for debugging purposes, however part of the desired output is that the resulting .parquet file(s) are ordered by the H3 index, so queries that filter by both H3 resolution 8 and H3 at original resolution fields are (hopefully) faster. E.g of queries that are intended to help with this new data organisation:

```sql
SELECT AVG({column}) FROM {table} WHERE h3_res8 >= {h3_res8_min} AND h3_res8 <= {h3_res8_max} AND h3_res11 >= {h3_res11_min} AND h3_res8 <= {h3_res11_max};
```


```sql
SELECT h3_res11, {column} FROM {table} WHERE h3_res8 >= {h3_res8_min} AND h3_res8 <= {h3_res8_max} AND h3_res11 >= {h3_res11_min} AND h3_res8 <= {h3_res11_max};
```

**Reply by Mmoncadaisla** [reply]
@Maxxen with this in mind, would you recommend directly going for the `COPY` query without first creating the table in the duckdb database? what do you think could be the cause for a "simple" CREATE TABLE AS SELECT * like query consuming so much memory when the source dataset ain't "that big" (specially in the last case where the sum of .parquet files size is ~50 GB)?
